flatpak manifest (org.flatpak.Gourmet.yml) and build instructions (flatpak.yaml)--also updated INSTALL.md

OzarkShepherd opened the pull request:
This flatpak manifest (org.flatpak.Gourmet.yml) works for me in Ubuntu 20.04, but the manifest should work in any distro.  The build instructions (flatpak.yaml) will vary by distro, but it works in Ubuntu 20.04.

cydanil commented on .github/workflows/flatpak.yaml:
```diff
@@ -0,0 +1,23 @@
+    
+    - name: install Flatpak
+      run: >
+        sudo apt install flatpak gnome-software-plugin-flatpak && flatpak remote-add --if-not-exists flathub https://flathub.org/repo/flathub.flatpakrepo && flatpak install flathub org.gnome.Platform//3.36 org.gnome.Sdk//3.36 org.flatpak.Builder
```
The two commands for `apt` and `flatpak` should be split on two different lines, there is no need to `&&` them

OzarkShepherd replied:
ok, thank you.  that is done now

cydanil commented on .github/workflows/flatpak.yaml:
```diff
@@ -0,0 +1,23 @@
+      run: >
+        flatpak-builder --run --socket=fallback-x11 build-dir org.flatpak.Gourmet.yml gourmet
+     
+# Until issue # 107 is fixed, in the directory Gourmet was built in, open the settings.py in the following path
```
When this is merged, could you add a comment to #107 to say that this file should be updated to remove these comments?

OzarkShepherd replied:
ok, that is done now, thank you

cydanil commented on org.flatpak.Gourmet.yml:
```diff
@@ -0,0 +1,54 @@
+app-id: org.flatpak.Gourmet
+runtime: org.gnome.Platform
+runtime-version: '3.36'
```
Could you shed some light on this number?

OzarkShepherd replied:
Flatpaks target a runtime, sdk, and version.  KDE, Gnome, and Freedesktop are the choices, along with their versions.  I went with Gnome 3.36 for the flatpak.

cydanil commented on org.flatpak.Gourmet.yml:
```diff
@@ -0,0 +1,54 @@
+
+  - name: gourmet
+    ensure-writable:
+      - /lib/python2.7/site-packages/easy-install.pth
```
`python3.8`?

OzarkShepherd replied:
Sorry, I started the flatpak for the original Gourmet, and forgot to remove that line when I switched to this fork for python 3.  It is removed now.

cydanil commented on .github/workflows/flatpak.yaml:
```diff
@@ -0,0 +1,23 @@
+jobs: 
+  build: flatpack
+  - uses: actions/checkout@v2   # automatically clone the repository
+    runs-on: ubuntu-20.04  # verified on ubuntu-20.04 and should run on any debian distro
```
In this file, the `runs-on` directive sets the platform to use for the workflow. The platforms available are [listed here](https://docs.github.com/en/actions/reference/software-installed-on-github-hosted-runners) . I think the comment can be removed.

OzarkShepherd replied:
ok, it is removed now, thank you.

cydanil commented on .github/workflows/flatpak.yaml:
```diff
@@ -0,0 +1,29 @@
+    - uses: actions/checkout@v2 
+    - name: Install Flatpak
+      run: >
+        sudo apt install flatpak gnome-software-plugin-flatpak && flatpak remote-add --if-not-exists flathub https://flathub.org/repo/flathub.flatpakrepo
```
Is `gnome-software-plugin-flatpak` required?

OzarkShepherd replied:
You are right, that is only needed for updating flatpaks through a GUI instead of manually through a terminal.  It isn't relevant for this manifest unless it gets submitted to flatpak for general use, so I removed it just now.

cydanil commented on .github/workflows/flatpak.yaml:
```diff
@@ -0,0 +1,30 @@
+    - name: Install Flatpak
+      run: >
+        sudo apt install flatpak && flatpak remote-add --if-not-exists flathub https://flathub.org/repo/flathub.flatpakrepo
+        flatpak install flathub org.gnome.Platform//3.36 org.gnome.Sdk//3.36 org.flatpak.Builder
```
This requires the `-y` flag, to accept installing the right flatpak dependencies (9 of them?)

OzarkShepherd replied:
The Platform and Sdk also have Locale dependencies, so that is 4; the other 5 was probably from having flatpak install the builder instead of apt.  The commands have been altered to take out the flatpak builder from flathub, and to use apt instead so there should be 4 dependencies in the flatpak install step now.

cydanil commented on .github/workflows/flatpak.yaml:
```diff
@@ -0,0 +1,30 @@
+    runs-on: ubuntu-20.04
+    
+    steps:
+    - uses: actions/checkout@v2 
```
Actually, this action is not needed, as `flatpack-builder` will clone the repo for us later

cydanil replied:
It is needed, to get the flatpak yaml config.

cydanil commented on .github/workflows/flatpak.yaml:
```diff
@@ -0,0 +1,30 @@
+    - uses: actions/checkout@v2 
+    - name: Install Flatpak
+      run: >
+        sudo apt install flatpak && flatpak remote-add --if-not-exists flathub https://flathub.org/repo/flathub.flatpakrepo
```
`flatpak-builder also needs to be installed:
```
sudo apt-get install flatpak flatpak-builder
```

OzarkShepherd replied:
I changed it to having apt install the flatpak-builder instead of flathub now.

cydanil commented:
I think that the manifest should be called `com.github.thinkle.Gourmet.yaml`, which is more consistent with other flatpak packages.

OzarkShepherd commented:
> I think that the manifest should be called `com.github.thinkle.Gourmet.yaml`, which is more consistent with other flatpak packages.

I got the name format from the flatpak documentation https://docs.flatpak.org/en/latest/first-build.html so I thought it was consistent with other flatpaks.  

If the flatpak stays only on github and never makes it to flatpak then I could see naming it com.github.kirienko.Gourmet.yaml to avoid confusion with the original version and to make it clear that it won't go on flatpak.  This flatpak will not work with the old Gourmet because too many dependencies are now deprecated or not even available.  The ones that are available tend to be unmaintained forks that I wasn't sure about anyway.  I did try to make a flatpak for that one first, after trying to contact the maintainer at launchpad for permission, but I failed to get it to work.  So to say that the flatpak is for the thinkle github is not correct in my opinion, since it only works for this kirienko fork.

cydanil commented on org.flatpak.Gourmet.yml:
```diff
@@ -0,0 +1,52 @@
+    buildsystem: simple
+    build-options:
+      build-args:
+        - --share=network
```
I think we would also need ` - --socket=pulseaudio` for the timer alerts

OzarkShepherd replied:
You are right, I am trying to figure out where it will work.  I tried it with where pyglet gets installed, but that didn't work just now.

OzarkShepherd replied:
got it, it works when added to the run command.  I'll adjust flatpak.yaml to enable the pulseaudio socket.

cydanil commented:
> > I think that the manifest should be called `com.github.thinkle.Gourmet.yaml`, which is more consistent with other flatpak packages.
> 
> I got the name format from the flatpak documentation https://docs.flatpak.org/en/latest/first-build.html so I thought it was consistent with other flatpaks.

That's true for the example. But the [documentation](https://docs.flatpak.org/en/latest/conventions.html#application-ids) prefers the `io.github` id. Since we don't have a homepage set up at the moment, I think that [minder](https://flathub.org/apps/details/com.github.phase1geo.minder) does it correctly.

> If the flatpak stays only on github and never makes it to flatpak then I could see naming it com.github.kirienko.Gourmet.yaml to avoid confusion with the original version and to make it clear that it won't go on flatpak. This flatpak will not work with the old Gourmet because too many dependencies are now deprecated or not even available. The ones that are available tend to be unmaintained forks that I wasn't sure about anyway. I did try to make a flatpak for that one first, after trying to contact the maintainer at launchpad for permission, but I failed to get it to work. So to say that the flatpak is for the thinkle github is not correct in my opinion, since it only works for this kirienko fork.

Yes! However, we aim to eventually bring all the changes upstream to [thinkle/gourmet](https://github.com/thinkle/gourmet). Thus, I think it would be less error-prone to name the file *seemingly* incorrectly for now :)

OzarkShepherd commented:
Ok, the flatpak manifest has been renamed to io.github.thinkle.Gourmet.yml

OzarkShepherd commented:
I also updated the INSTALL.md but did not realize that it would go into the same pull request.  In the INSTALL.md I took out parts about Gourmet being prepackaged, changed the python to 3.8 and changed the dependencies, mentioned that this is still in development, and included a new section specifically for Ubuntu 20.04.

OzarkShepherd commented:
In the INSTALL.md, pulling in all the dependencies also adds unexpected packages like the chromium snap and some drivers.  I don't know why it does that.

cydanil commented:
This is probably due to the mention of `selenium`, which mention I'm unsure of. `selenium` is an GUI automation framework, and we shouldn't need it.

While I appreciate that you're updating `INSTALL.md`, PR should be as small as possible, and focus on a single issue.
I suggest that you split the PR into two: one for the flatpak, and one for the documentation update.

With regards with the Flatpak, I think that:
- [ ] There should be a [name](https://docs.github.com/en/actions/reference/workflow-syntax-for-github-actions#name) to the workflow (eg. `Create Flatpak`);
- [ ] `workflow_dispatch` does not need inputs;
- [ ] The flatpak should be packed into a single file (`gourmet.flatpak`) that can be made available for download.
- [ ] What about internationalization? Although I suppose it can be left out for a future PR.

The last point is tricky: the build dir is cannot be compressed directly, it seems. Rather, `flatpak build-bundle` requires the files to be available from some repo. Is that what `flatpak build-export` does? Or is it `build-update-repo`?

OzarkShepherd commented:
-Ok, I took out selenium from the flatpak manifest and it still seems to work.  It should also be taken out of https://github.com/kirienko/gourmet/blob/master/requirements.txt 
-I also took out the workflow inputs and added a name to the flatpak.yaml
-I will look into how to compress the build dir into a flatpak file
-I thought elib.intl doesn't work anymore and don't know what to replace it with. There are other missing plugins but I don't know what packages to add for the plugins.
-I'll ask on slack about how to split the pull request, because it looks complicated from the google search I did.

OzarkShepherd commented:
I got a single compressed directory into a tar, which contains the repo.  Size is 110 MB and does not include the Gnome 3.36 runtime which has to be installed separately.  I could put instructions inside the tar if they would be useful.  I could see this being useful for end users, but for testing it would have to be rebuilt each time a change is made.  Also app launcher icons are not being installed at this time through this tar, which I believe is a sandbox permission problem during the building so it should just take some research to fix.  I am hesitant to upload the new tar to my fork right now though because I don't want to make this PR more complicated.

OzarkShepherd commented:
the changes to INSTALL.md should now be reverted to the original before I modified it.

OzarkShepherd commented:
I am closing this PR so I can reset my fork and make branches.  Then I can make smaller pull requests.
